### PR TITLE
Helm exporter to add the operator as dep to charts

### PR DESCRIPTION
**Pull request by zeenix**
From the last commit's log:

    Add dependency to the Habitat operator chart in the generated charts. This
    automates the installation of the operator into the Kubernetes cluster and
    is installed as part of the generated chart for the user.
    
    Unless '--download-deps' CLI option is passed to the exporter, user of the
    exporter has to manually run
    
    $ helm repo add habitat-operator \
      https://kinvolk.github.io/habitat-operator/helm/charts/stable/
    $ helm dep up PATH_OF_GENERATED_CHART_DIR
    
    before they can install the chart or package it.
    
    Before the first time the exporter is run, Helm needs to be setup, which
    is just a matter of `helm init -c` if one needs to setup Helm for the
    purpose of creating charts only and not actually installing them on a
    Kubernetes cluster.

### CODE REVIEW

File: components/pkg-export-helm/src/deps.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+use clap;
+use std::io::Write;
+use std::path::Path;
+use std::process::Command;
+
+use common::ui::{UI, Status};
+use export_docker::Result;
+
+// Keep the default operator version in main::cli() in sync with this one
+pub const DEFAULT_OPERATOR_VERSION: &'static str = "0.4.0";
```
Comment (lilic) [inline components/pkg-export-helm/src/deps.rs]: Does this mean that each time we will release up a new version of habitat operator that needs to be changed here?

File: components/pkg-export-helm/src/main.rs [hunk 2]
```diff
@@ -95,6 +95,17 @@ fn cli<'a, 'b>() -> clap::App<'a, 'b> {
                 .long("desc")
                 .help("A single-sentence description"),
         )
+        .arg(
+            Arg::with_name("OPERATOR_VERSION")
+                .value_name("OPERATOR_VERSION")
+                .long("operator-version")
+                .validator(valid_version)
+                // Keep the default version here in sync with deps::DEFAULT_OPERATOR_VERSION
+                .help("Version of the Habitat operator to set as dependency (default: 0.4.0)"),
```
Comment (lilic) [inline components/pkg-export-helm/src/main.rs]: Can't you just then use the `deps::DEFAULT_OPERATOR_VERSION` here?

File: components/pkg-export-helm/src/deps.rs (hunk 1 above)
Comment (lilic) [reply components/pkg-export-helm/src/deps.rs]: Also now that we have the version we can put it in the CLI ;)

**Review by lilic — CHANGES_REQUESTED** [review]
Some questions.

File: components/pkg-export-helm/src/deps.rs (hunk 1 above)
Comment (zeenix) [reply components/pkg-export-helm/src/deps.rs]: Yes. I'll be happy to consider any possible alternatives cause I don't like it either.

File: components/pkg-export-helm/src/main.rs (hunk 2 above)
Comment (zeenix) [reply components/pkg-export-helm/src/main.rs]: Unfortunately not. `help` method expects a static string and I was informed on the IRC that currently there is no way of concetanating two static strings into one but now I googled again and found a crate that might make it possible:

https://github.com/rust-lang-nursery/lazy-static.rs

which seems to be even used by other parts of Habitat so I'll give it a go (or rust :-P).

Comment (zeenix) [reply components/pkg-export-helm/src/main.rs]: It works! :)

**Comment by zeenix** [comment]
(forced) pushed changes. Please check again.

File: components/pkg-export-helm/src/main.rs [hunk 3]
```diff
@@ -84,17 +98,27 @@ fn cli<'a, 'b>() -> clap::App<'a, 'b> {
...
         )
+        .arg(
+            Arg::with_name("OPERATOR_VERSION")
+                .value_name("OPERATOR_VERSION")
+                .long("operator-version")
+                .validator(valid_version)
+                .help(&OPERATOR_VERSION_HELP),
+        )
+        .arg(Arg::with_name("DOWNLOAD_DEPS").long("download-deps").help(
+            "Whether to download dependencies (default: no)",
```
Comment (lilic) [inline components/pkg-export-helm/src/main.rs]: > Whether to download dependencies (default: no)

Think that could use rephrasing? Also maybe explain what kind of dependencies here, especially for new Helm users.

File: components/pkg-export-helm/src/main.rs [hunk 4]
```diff
@@ -41,6 +44,18 @@ use hcore::PROGRAM_NAME;
...
 use chart::Chart;
 
+lazy_static! {
+    static ref CHART_VERSION_HELP: String = format!(
+        "Version of the chart to create (default: {})",
+        chartfile::DEFAULT_VERSION,
+    );
+
+    static ref OPERATOR_VERSION_HELP: String = format!(
+        "Version of the Habitat operator to set as dependency (default: {})",
```
Comment (lilic) [inline components/pkg-export-helm/src/main.rs]: Can we add this now the k8s exporter CLI as well please :) (not part of this PR of course)

File: components/pkg-export-helm/src/main.rs [hunk 5]
```diff
@@ -63,8 +78,7 @@ fn export_for_cli_matches(ui: &mut UI, matches: &clap::ArgMatches) -> Result<()>
 fn cli<'a, 'b>() -> clap::App<'a, 'b> {
     let name: &str = &*PROGRAM_NAME;
     let about = "Creates a Docker image and generates a Helm chart for the specified Habitat \
-                 package. Habitat operator must be deployed within the Kubernetes cluster before \
-                 the generated chart can be installed.";
+                 package.";
```
Comment (lilic) [inline components/pkg-export-helm/src/main.rs:67]: If the default for downloading dependencies is no, then we should still leave this msg here, no?

**Review by lilic — CHANGES_REQUESTED** [review]
some more comments...

File: components/pkg-export-helm/src/deps.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+use handlebars::Handlebars;
+use clap;
+use std::io::Write;
+use std::path::Path;
+use std::process::Command;
+
+use common::ui::{UI, Status};
+use export_docker::Result;
+
+pub const DEFAULT_OPERATOR_VERSION: &'static str = "0.4.0";
```
Comment (asymmetric) [inline components/pkg-export-helm/src/deps.rs]: You can update this now.

File: components/pkg-export-helm/src/deps.rs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        Ok(())
+    }
+
+    pub fn download<P: AsRef<Path>>(&self, dir: P, ui: &mut UI) -> Result<()> {
+        if !self.update {
+            return Ok(());
+        }
+        ui.status(Status::Downloading, "dependencies")?;
+
+        // Ignore the results cause it's not a problem if this execution fails.
```
Comment (asymmetric) [inline components/pkg-export-helm/src/deps.rs]: Why not? Maybe add it to the comment.

File: components/pkg-export-helm/src/main.rs (hunk 5 above)
Comment (zeenix) [reply components/pkg-export-helm/src/main.rs:67]: no, the dep is always added so after export, user has to run `helm dup up` anyway *before* packaging/distributing the chart. i-e No operator required to be in the cluster.

File: components/pkg-export-helm/src/main.rs (hunk 3 above)
Comment (asymmetric) [reply components/pkg-export-helm/src/main.rs]: You can use [`default_value`](https://docs.rs/clap/2.30.0/clap/struct.Arg.html#method.default_value) to set defaults in clap.

**Comment by zeenix** [comment]
PTALA. This PR now also include the needed k8s update wrt to latest operator changes.

File: components/pkg-export-kubernetes/src/manifest.rs [hunk 8]
```diff
@@ -31,8 +31,12 @@ use topology::Topology;
 /// Represents a Kubernetes manifest.
 #[derive(Debug, Clone)]
 pub struct Manifest {
+    /// The identifier of the Habitat package
+    pub pkg_ident: PackageIdent,
     /// The Habitat service name.
     pub metadata_name: String,
+    /// Name of the Kubernetes resource.
```
Comment (lilic) [inline components/pkg-export-kubernetes/src/manifest.rs]: what is this `k8s_name`, not sure I understand completely...

File: components/pkg-export-helm/src/deps.rs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+use clap;
+use std::io::Write;
+use std::path::Path;
+use std::process::Command;
+
+use common::ui::{UI, Status};
+use export_docker::Result;
+
+pub const DEFAULT_OPERATOR_VERSION: &'static str = "0.5.1";
+pub const OPERATOR_REPO_URL: &'static str = "https://kinvolk.github.io/habitat-operator\
```
Comment (krnowak) [inline components/pkg-export-helm/src/deps.rs:26]: Should this also be configurable via command line options?

File: components/pkg-export-kubernetes/tests/KubernetesManifestTestBinds.yaml [hunk 10]
```diff
@@ -12,6 +12,8 @@ spec:
   ## An object containing parameters that affects how the Habitat service
   ## is executed.
   service:
+    ## Name of the Habitat service.
+    name: nginx-latest
```
Comment (lilic) [inline components/pkg-export-kubernetes/tests/KubernetesManifestTestBinds.yaml]: don't think this is the correct name of the service... more like `nginx`.

**Review by lilic — CHANGES_REQUESTED** [review]

File: components/pkg-export-helm/src/chart.rs [hunk 11]
```diff
@@ -109,79 +117,91 @@ impl<'a> Chart<'a> {
...
         self.ui.status(
             Status::Creating,
-            format!("manifest template `{}`", manifest_path),
+            format!("file `{}`", manifest_path),
         )?;
         let mut write = fs::File::create(manifest_path)?;
-        let out: String = self.manifest_template.into();
+        let out: String = self.manifest_template
+            .take()
+            .expect("generate_manifest_template() called more than once")
```
Comment (krnowak) [inline components/pkg-export-helm/src/chart.rs:171]: This is kinda icky, but I guess solving this in a build-time safe manner would be a bit more involved (likely a separate type with similar functions like Chart has, but without the `generate_manifest_template` one. Also, it's a private function, so I guess we are fine.

**Review by krnowak — COMMENTED** [review]
Looks good, one question basically.

File: components/pkg-export-kubernetes/src/manifest.rs (hunk 8 above)
Comment (zeenix) [reply components/pkg-export-kubernetes/src/manifest.rs]: that will be the name used for the k8s resource IIUC. It's what will now map to metadata/name in the manifest.

File: components/pkg-export-helm/src/deps.rs (hunk 9 above)
Comment (zeenix) [reply components/pkg-export-helm/src/deps.rs:26]: maybe but not so important IMO since it can be set as part of `helm install..` command.

Comment (zeenix) [reply components/pkg-export-helm/src/deps.rs:26]: actually, no this part can't be configured afterwards. However, let's make them configurable later.

Comment (krnowak) [reply components/pkg-export-helm/src/deps.rs:26]: Alright.

File: components/pkg-export-kubernetes/tests/KubernetesManifestTestBinds.yaml (hunk 10 above)
Comment (zeenix) [reply components/pkg-export-kubernetes/tests/KubernetesManifestTestBinds.yaml]: The `-latest` suffix comes from the time when we agreed to keep the (docker) image name and metadata/name the same for keeping it simple. I'm guessing this new property requires the name to be the same as (habitat) pacakge name? When the version and release info is available, do we want this to be `nginx-VERSION-RELEASE`?

Comment (lilic) [reply components/pkg-export-kubernetes/tests/KubernetesManifestTestBinds.yaml]: no, we want it to the be the same as the habitat service/pkg name. In the case of nginx that is nginx. See https://github.com/kinvolk/habitat-operator/pull/155 for more info on this change.

Comment (zeenix) [reply components/pkg-export-kubernetes/tests/KubernetesManifestTestBinds.yaml]: The PR description says:

> This also allows users to use any name they want for the ObjectMeta.Name field, e.g. myapp-v1, making it possible to run multiple versions of the same Service.

To make the last part work (without user intervention), we should be adding the -VERSION-RELEASE here, shouldn't we?

Comment (lilic) [reply components/pkg-export-kubernetes/tests/KubernetesManifestTestBinds.yaml]: You are mixing up the Service name and the Metadata name... the service name should NOT include anything but the service name.

**Comment by zeenix** [comment]
PTALA.

File: components/pkg-export-helm/src/chartfile.rs [hunk 12]
```diff
@@ -17,7 +17,6 @@ use handlebars::Handlebars;
 
 use export_docker::Result;
 
-// Keep the default version in main::cli() in sync with this one
 pub const DEFAULT_VERSION: &'static str = "0.0.1";
```
Comment (lilic) [inline components/pkg-export-helm/src/chartfile.rs:20]: What is this default version of?

File: components/pkg-export-helm/src/main.rs [hunk 13]
```diff
@@ -41,6 +42,7 @@ use hcore::PROGRAM_NAME;
 
 use chart::Chart;
 
+
```
Comment (lilic) [inline components/pkg-export-helm/src/main.rs]: How come this empty extra line? :)

File: components/pkg-export-helm/src/main.rs [hunk 14]
```diff
@@ -84,17 +85,30 @@ fn cli<'a, 'b>() -> clap::App<'a, 'b> {
         .arg(
             Arg::with_name("VERSION")
                 .value_name("VERSION")
+                .short("v")
                 .long("version")
                 .validator(valid_version)
-                // Keep the default version here in sync with chartfile::DEFAULT_VERSION
-                .help("Version of the chart to create (default: 0.0.1)"),
+                .help("Version of the chart to create")
```
Comment (lilic) [inline components/pkg-export-helm/src/main.rs:90]: Aha nvm my question above. :D

File: components/pkg-export-helm/src/main.rs [hunk 15]
```diff
@@ -84,17 +85,30 @@ fn cli<'a, 'b>() -> clap::App<'a, 'b> {
...
+        .arg(
+            Arg::with_name("OPERATOR_VERSION")
+                .value_name("OPERATOR_VERSION")
+                .long("operator-version")
+                .validator(valid_version)
+                .help("Version of the Habitat operator to set as dependency")
+                .default_value(deps::DEFAULT_OPERATOR_VERSION),
+        )
+        .arg(Arg::with_name("DOWNLOAD_DEPS").long("download-deps").help(
+            "Whether to download dependencies. Currently only the Habitat Kubernetes Operator \
```
Comment (lilic) [inline components/pkg-export-helm/src/main.rs]: Nit: a bit weird phrasing "Currently only the Habitat Kubernetes Operator". maybe "The Kubernetes habitat-operator is the only dependancy currently."

File: components/pkg-export-kubernetes/defaults/KubernetesManifest.hbs [hunk 16]
```diff
@@ -24,6 +24,8 @@ spec:
   ## An object containing parameters that affects how the Habitat service
   ## is executed.
   service:
+    ## Name of the Habitat service.
```
Comment (lilic) [inline components/pkg-export-kubernetes/defaults/KubernetesManifest.hbs]: hmm maybe Habitat pkg would make more sense to the Habitat users. Maybe leave it like this, and I can bring it up in the upstream PR and we can see what they think makes more sense to them.

File: components/pkg-export-kubernetes/src/cli.rs [hunk 17]
```diff
@@ -91,6 +91,12 @@ impl<'a, 'b> Cli<'a, 'b> {
     pub fn add_runtime_args(self) -> Self {
         Cli {
             app: self.app
+                .arg(
+                    Arg::with_name("K8S_NAME")
+                        .value_name("K8S_NAME")
+                        .long("k8s-name")
+                        .help("The Kubernetes resource name"),
```
Comment (lilic) [inline components/pkg-export-kubernetes/src/cli.rs]: Maybe like you should mention the default is pkg_name + version? WDYT?

**Review by lilic — APPROVED** [review]
A few nits and some suggestions, otherwise feel free to submit upstream.

File: components/pkg-export-kubernetes/src/cli.rs (hunk 17 above)
Comment (zeenix) [reply components/pkg-export-kubernetes/src/cli.rs]: Sure

File: components/pkg-export-helm/src/main.rs (hunk 15 above)
Comment (zeenix) [reply components/pkg-export-helm/src/main.rs]: ok

File: components/pkg-export-helm/src/main.rs (hunk 13 above)
Comment (zeenix) [reply components/pkg-export-helm/src/main.rs]: typo i guess..

**Comment by zeenix** [comment]
nits addressed I think.
